src/main/java/nl/tudelft/lifetiles/graph/model/BucketCache.java

```diff
@@ -0,0 +1,155 @@
+package nl.tudelft.lifetiles.graph.model;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import nl.tudelft.lifetiles.sequence.model.SequenceSegment;
+
+/**
+ * A bucket cache which adds the sequence segments to buckets for improved
+ * graph drawing performance.
+ *
+ * @author Jos
+ *
+ */
+public class BucketCache {
+
+    /**
+     * Graph that has been inserted into the bucket cache.
+     */
+    private final Graph<SequenceSegment> graph;
+
+    /**
+     * Number of buckets the graph is divided in.
+     */
+    private final int numberBuckets;
+
+    /**
+     * Buckets used to store the sequenceSegments.
+     */
+    private List<SortedSet<SequenceSegment>> buckets;
+
+    /**
+     * Width of a bucket based on the width of the graph and the number of
+     * buckets being cached.
+     */
+    private final long bucketWidth;
+
+    /**
+     * Max unified end coordinate position in the graph, should be stored
+     * because it is called on every view update.
+     */
+    private final long maxUnifiedEnd;
+
+    /**
+     * Constructs a bucket cache and divides the graph into n buckets.
+     *
+     * @param numberBuckets
+     *            Number of buckets the graph is divided in.
+     * @param graph
+     *            Graph that needs to be divided.
+     */
+    public BucketCache(final int numberBuckets,
+            final Graph<SequenceSegment> graph) {
+        this.numberBuckets = numberBuckets;
+        this.graph = graph;
+        maxUnifiedEnd = getMaxUnifiedEnd();
+        bucketWidth = maxUnifiedEnd / this.numberBuckets + 1;
+        cacheGraph();
+    }
+
+    /**
+     * Caches the graph that has been inserted into the bucket cache into the
+     * number of buckets the graph should be divided in.
+     */
+    private void cacheGraph() {
+        long startTime = Calendar.getInstance().getTimeInMillis();
+        buckets = new ArrayList<SortedSet<SequenceSegment>>();
+        for (int index = 0; index < numberBuckets; index++) {
+            buckets.add(index, new TreeSet<SequenceSegment>());
+        }
+        for (SequenceSegment vertex : graph.getAllVertices()) {
+            cacheVertex(vertex);
+        }
+        System.out.println("Graph caching. Took "
+                + (Calendar.getInstance().getTimeInMillis() - startTime)
+                + " ms.");
+    }
+
+    /**
+     * Caches the vertex to the bucket cache, a vertex can be in multiple
+     * buckets in the bucket cache.
+     *
+     * @param vertex
+     *            The vertex to cache in the buckets in the cache.
+     */
+    private void cacheVertex(final SequenceSegment vertex) {
+        int startBucket = (int) Math.floor((double) vertex.getUnifiedStart()
```

Comment:
You should modify the method `bucketPosition` with an additional argument so that you can use it here too